tenseal/tensealencoder.h

```diff
@@ -25,6 +26,65 @@ class TenSEALEncoder {
         return std::any_cast<std::shared_ptr<T>>(_encoders[tidx]);
     }
 
+    /*
+    Template encoding functions to choose between the use of
+    Integer/BatchEncoder or CKKSEncoder.
+    */
+    template <class T>
+    void encode(std::vector<int64_t>& vec, Plaintext& pt) {
+        auto encoder = this->get<T>();
+        encoder->encode(vec, pt);
+    }
+
+    template <class CKKSEncoder>
+    void encode(std::vector<double>& vec, Plaintext& pt,
+                std::optional<double> optscale = {}) {
+        double scale = 1.0;
+        if (optscale.has_value())
+            scale = optscale.value();
+        else
+            scale = global_scale();
+
+        auto encoder = this->get<CKKSEncoder>();
+        encoder->encode(vec, scale, pt);
+    }
+
+    /*
+    Template decoding functions Integer/BatchEncoder/CKKSEncoder.
+    */
+    template <class T, class R>
+    void decode(const Plaintext& pt, R& result) {
+        auto encoder = this->get<T>();
+        encoder->decode(pt, result);
+    }
+
+    /*
+    Template for slot count.
+    */
+    template <class T>
+    size_t slot_count() {
+        auto encoder = this->get<T>();
+        return encoder->slot_count();
+    }
+
+    // throws invalid_argument exception if the scale is incompatible with
+    // current primes used. Example: if using coeff_mod_bit_size of
+    // [60,40,40,60], the global scale should be set for 2**40
+    void global_scale(double scale) {
+        auto parms_id = _context->first_parms_id();
+        auto ctx_data = _context->get_context_data(parms_id);
+        if (static_cast<int>(log2(scale)) >=
+            ctx_data->total_coeff_modulus_bit_count())
+            throw std::invalid_argument("invalid scale value");
```

Comment:
I don't think we should do anything about it at this point, errors should be forwarded to SEAL

Reply:
we can maybe force it to be positive here...

Reply:
ok. I was following the commented TODO.

SEAL throws an exception anyway, from a similar condition.

Reply:
Ah, the TODO was about checking if the scale matches the coeff[1], as you may know, rescaling is dividing by each prime, and if we use primes with bit_size different than the scale, then we will probably end with a bad decryption. We shouldn't catch error that are checked by SEAL but rather focus on specific ones for TenSEAL.

Don't bother yourself with the TODO, a full strategy for checking those value need to be designed first, there is a lot to think of, and we can defer this to a later stage.